### PR TITLE
!!! TASK: Clean up password hashing strategies

### PR DESCRIPTION
**Pull request by kdambekalns**
Cleans up the password hashing strategies shipped with Flow.

**Upgrade instructions**

If you implemented `PasswordHashingStrategyInterface`, take note that type declarations will be added for the next major version (9.0) so adjust your implementation to use the added type declarations like in the core implementations.

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [x] Reviewer - The first section explains the change briefly for change-logs
- [x] Reviewer - Breaking Changes are marked wit `!!!` and have upgrade-instructions

### CODE REVIEW

**Comment by kdambekalns** [comment]
`password_hash()` returns strings with `$2y$` instead of "our" `$2a$` inside. But `password_verify()` decodes "our" hashed BCrypt passwords fine, see https://3v4l.org/L0QnW#v8.1.11

**Comment by kdambekalns** [comment]
Note: We can of course drop the interface change (or move it to 9.0), but I'd say the risk is low. Who would implement their own hashing strategy? 😇

**Comment by kdambekalns** [comment]
Ping…

**Review by albe — APPROVED** [review]

**Comment by kdambekalns** [comment]
@markusguenther (and others), any objections to the interface change?

**Review by fcool — APPROVED** [review]

**Comment by fcool** [comment]
Ah... hmmm... 
Strictly speaking this would be a breaking change, which might be not appropriate for 8.2 - at least without REALLY good reasons...

What about NOT changing the interface in 8.2 - but do so in 9.0? (Rest of the code can stay as it is, if I see it correctly, as "additional hints" are allowed)

On the other hand... I cannot believe, that this interface has been implemented too often outside the Flow packages themself. And it would be quite easy to be forward compatible. (So I would not say, we absolute cannot merge that into 8.2)

**Comment by bwaidelich** [comment]
> Strictly speaking this would be a breaking change, which might be not appropriate for 8.2 - at least without REALLY good reasons...

I agree

File: Neos.Flow/Classes/Security/Cryptography/PasswordHashingStrategyInterface.php [hunk 1]
```diff
@@ -21,18 +23,18 @@ interface PasswordHashingStrategyInterface
      * Hash a password for storage
      *
      * @param string $password Cleartext password that will be hashed
-     * @param string $staticSalt Optional static salt that will not be stored in the hashed password
+     * @param string|null $staticSalt Optional static salt that will not be stored in the hashed password
      * @return string The hashed password with dynamic salt (if used)
      */
-    public function hashPassword($password, $staticSalt = null);
+    public function hashPassword(string $password, string $staticSalt = null): string;
```
Comment (kitsunet) [inline Neos.Flow/Classes/Security/Cryptography/PasswordHashingStrategyInterface.php]:
```suggestion
    public function hashPassword($password, $staticSalt = null);
```

File: Neos.Flow/Classes/Security/Cryptography/PasswordHashingStrategyInterface.php [hunk 2]
```diff
@@ -21,18 +23,18 @@ interface PasswordHashingStrategyInterface
...
      *
      * @param string $password
      * @param string $hashedPasswordAndSalt Hashed password with dynamic salt (if used)
-     * @param string $staticSalt Optional static salt that will not be stored in the hashed password
-     * @return boolean true if the given cleartext password matched the hashed password
+     * @param string|null $staticSalt Optional static salt that will not be stored in the hashed password
+     * @return bool true if the given cleartext password matched the hashed password
      */
-    public function validatePassword($password, $hashedPasswordAndSalt, $staticSalt = null);
+    public function validatePassword(string $password, string $hashedPasswordAndSalt, string $staticSalt = null): bool;
```
Comment (kitsunet) [inline Neos.Flow/Classes/Security/Cryptography/PasswordHashingStrategyInterface.php]:
```suggestion
    public function validatePassword($password, $hashedPasswordAndSalt, $staticSalt = null);
```

**Review by kitsunet — CHANGES_REQUESTED** [review]
Agreed with merging ASAP but without the Interface change.

**Review by jonnitto — APPROVED** [review]
Fine by reading

**Comment by kdambekalns** [comment]
> (Rest of the code can stay as it is, if I see it correctly, as "additional hints" are allowed)

No, as the test results show…

File: Neos.Flow/Classes/Security/Cryptography/BCryptHashingStrategy.php [hunk 3]
```diff
@@ -11,45 +13,41 @@
...
      *
-     * @param string $password   The plaintext password to hash
-     * @param string $staticSalt Optional static salt that will not be stored in the hashed password
-     * @return string the result of the crypt() call
+     * @param string $password The plaintext password to hash
+     * @param string|null $staticSalt Not used with this strategy
+     * @return string The hashed password
      */
-    public function hashPassword($password, $staticSalt = null)
+    public function hashPassword(string $password, string $staticSalt = null): string
```
Comment (kdambekalns) [inline Neos.Flow/Classes/Security/Cryptography/BCryptHashingStrategy.php]:
```suggestion
    public function hashPassword($password, $staticSalt = null)
```

File: Neos.Flow/Classes/Security/Cryptography/BCryptHashingStrategy.php [hunk 4]
```diff
@@ -59,17 +57,12 @@ public function hashPassword($password, $staticSalt = null)
...
      *
      * @param string $password The cleartext password
-     * @param string $hashedPasswordAndSalt The derived key and salt in as returned by crypt() for verification
-     * @param string $staticSalt Optional static salt that will be appended to the dynamic salt
+     * @param string $hashedPasswordAndSalt The derived key and salt in as returned by hashPassword() for verification
+     * @param string|null $staticSalt Not used with this strategy
      * @return boolean true if the given password matches the hashed password
      */
-    public function validatePassword($password, $hashedPasswordAndSalt, $staticSalt = null)
+    public function validatePassword(string $password, string $hashedPasswordAndSalt, string $staticSalt = null): bool
```
Comment (kdambekalns) [inline Neos.Flow/Classes/Security/Cryptography/BCryptHashingStrategy.php]:
```suggestion
    public function validatePassword($password, $hashedPasswordAndSalt, $staticSalt = null)
```

File: Neos.Flow/Classes/Security/Cryptography/Pbkdf2HashingStrategy.php [hunk 5]
```diff
@@ -64,10 +62,10 @@ public function __construct($dynamicSaltLength, $iterationCount, $derivedKeyLeng
      * Will use a combination of a random dynamic salt and the given static salt.
      *
      * @param string $password Cleartext password that should be hashed
-     * @param string $staticSalt Static salt that will be appended to the random dynamic salt
+     * @param string|null $staticSalt Static salt that will be appended to the random dynamic salt
      * @return string A Base64 encoded string with the derived key (hashed password) and dynamic salt
      */
-    public function hashPassword($password, $staticSalt = null)
+    public function hashPassword(string $password, string $staticSalt = null): string
```
Comment (kdambekalns) [inline Neos.Flow/Classes/Security/Cryptography/Pbkdf2HashingStrategy.php]:
```suggestion
    public function hashPassword($password, $staticSalt = null)
```

File: Neos.Flow/Classes/Security/Cryptography/Pbkdf2HashingStrategy.php [hunk 6]
```diff
@@ -80,11 +78,11 @@ public function hashPassword($password, $staticSalt = null)
...
      * @param string $password The cleartext password
      * @param string $hashedPasswordAndSalt The derived key and salt in Base64 encoding as returned by hashPassword for verification
-     * @param string $staticSalt Static salt that will be appended to the dynamic salt
-     * @return boolean true if the given password matches the hashed password
+     * @param string|null $staticSalt Static salt that will be appended to the dynamic salt
+     * @return bool true if the given password matches the hashed password
      * @throws \InvalidArgumentException
      */
-    public function validatePassword($password, $hashedPasswordAndSalt, $staticSalt = null)
+    public function validatePassword(string $password, string $hashedPasswordAndSalt, string $staticSalt = null): bool
```
Comment (kdambekalns) [inline Neos.Flow/Classes/Security/Cryptography/Pbkdf2HashingStrategy.php]:
```suggestion
    public function validatePassword($password, $hashedPasswordAndSalt, $staticSalt = null)
```

File: Neos.Flow/Tests/Unit/Security/Cryptography/Fixture/TestHashingStrategy.php [hunk 7]
```diff
@@ -15,24 +17,23 @@
...
 {
-
     /**
      * @param string $password Cleartext password that will be hashed
-     * @param string $staticSalt Optional static salt that will not be stored in the hashed password
+     * @param string|null $staticSalt Optional static salt that will not be stored in the hashed password
      * @return string The hashed password with dynamic salt (if used)
      */
-    public function hashPassword($password, $staticSalt = null)
+    public function hashPassword(string $password, string $staticSalt = null): string
```
Comment (kdambekalns) [inline Neos.Flow/Tests/Unit/Security/Cryptography/Fixture/TestHashingStrategy.php]:
```suggestion
    public function hashPassword($password, $staticSalt = null)
```

File: Neos.Flow/Tests/Unit/Security/Cryptography/Fixture/TestHashingStrategy.php [hunk 8]
```diff
@@ -15,24 +17,23 @@
...
 
     /**
      * @param string $password
      * @param string $hashedPasswordAndSalt Hashed password with dynamic salt (if used)
-     * @param string $staticSalt Optional static salt that will not be stored in the hashed password
+     * @param string|null $staticSalt Optional static salt that will not be stored in the hashed password
      * @return boolean true if the given cleartext password matched the hashed password
      */
-    public function validatePassword($password, $hashedPasswordAndSalt, $staticSalt = null)
+    public function validatePassword(string $password, string $hashedPasswordAndSalt, string $staticSalt = null): bool
```
Comment (kdambekalns) [inline Neos.Flow/Tests/Unit/Security/Cryptography/Fixture/TestHashingStrategy.php]:
```suggestion
    public function validatePassword($password, $hashedPasswordAndSalt, $staticSalt = null)
```

**Review by kdambekalns — COMMENTED** [review]
Revert remaining changes to methods from interface

**Comment by kitsunet** [comment]
Unfortunate :/ But then lets merge this as is now.

**Review by kitsunet — APPROVED** [review]

**Pull request merged**